test/Test/Database/LSMTree/Internal/Vector/Growing.hs

```diff
@@ -0,0 +1,212 @@
+            -> [Segment a] -- Segments
+            -> Vector a    -- Final vector
+finalVector initialBufferSize segments = runST $ do
+    vec <- new initialBufferSize
+    mapM_ (appendSegment vec) segments
+    freeze vec
+
+{-
+    Supplies the final contents of a growing vector for constructing a property.
+
+    The resulting property additionally provides statistics about the buffer
+    size scaling exponents, where a buffer size scaling exponent is the binary
+    logarithm of the ratio between the buffer size after and before appending a
+    segment.
+-}
+withFinalVector
+    :: Testable prop
+    => Int                -- Initial buffer size
+    -> [Segment a]        -- Segments
+    -> (Vector a -> prop) -- Property from final vector
+    -> Property           -- Resulting property
+withFinalVector initialBufferSize segments consumer
+    = tabulate
+          "Buffer size scaling exponents"
+          (show <$> bufferSizeScalingExponents availableBufferSizes 0 segments)
+          (consumer $ finalVector initialBufferSize segments)
+    where
+
+    availableBufferSizes :: [Int]
+    availableBufferSizes = iterate (* 2) initialBufferSize
+    {-
+        We do not need to account for overflow here, because the lengths of the
+        vectors we construct are small compared to @maxBound :: Int@.
+    -}
+
+    bufferSizeScalingExponents :: [Int] -> Int -> [Segment a] -> [Int]
+    bufferSizeScalingExponents _ _ []
+        = []
+    bufferSizeScalingExponents availableSizes length (head : tail)
+        = List.length insufficient :
+          bufferSizeScalingExponents sufficient length' tail
+        where
+
+        length' :: Int
+        !length' = length + segmentLength head
+
+        insufficient, sufficient :: [Int]
+        (insufficient, sufficient) = span (< length') availableSizes
+
+-- * Properties to test
+
+prop_finalVectorIsCorrect :: Positive (Small Int)
+                          -> Segments Integer
+                          -> Property
+prop_finalVectorIsCorrect (Positive (Small initialBufferSize))
+                          (Segments segments)
+    = withFinalVector initialBufferSize segments $ \ vec ->
+      toList vec === concatMap segmentToList segments
+
+-- * Test case generation and shrinking
+
+generateSegment :: Arbitrary a => Int -> Gen (Segment a)
+generateSegment maxCount = Replicate <$> chooseInt (0, maxCount) <*> arbitrary
+
+shrinkSegment :: Arbitrary a => Segment a -> [Segment a]
+shrinkSegment (Replicate count val)
+    = [Replicate count' val  | count' <- shrinkNat count] ++
+      [Replicate count  val' | val'   <- shrink val]
+    where
+
+    shrinkNat :: Int -> [Int]
+    shrinkNat = shrinkMap getNonNegative NonNegative
+
+{-
+    A list of segments to be appended to an initially empty vector. 'Segments a'
+    is isomorphic to '[Segment a]' but has a special way of generating test
+    cases, which avoids skewing with respect to buffer size scaling exponents.
```

Comment:
You could use `coverTable` or something similar to check that we are indeed growing vectors in the tests. It will also make the tests more future proof, because if anything about the test (distribution) changes, then the coverage might change, and then `coverTable` could catch that

Reply:
Well, the tabulation introduced by `withFinalVector` should already achieve this: no growth results in a buffer size scaling exponent of zero, and growth results in a buffer size scaling exponent greater than zero (and in the latter case the precise value even tells you how much the buffer is grown). Am I missing something?

Reply:
The point of `coverTable` is that you specify how often you expect each label to appear, and fail the test if the probability of a certain label is too small. So it's not just printing the labels so that we can look at them and see indeed that buffer size scaling exponents are non-zero, it's also asserting something about their distribution

Reply:
I’ve made a corresponding change in a83d9bbbfc0e44175774a27ca49afe4aade996cc.